core/src/main/java/org/apache/brooklyn/core/network/AbstractOnNetworkEnricher.java

```diff
@@ -0,0 +1,407 @@
+                    try {
+                        tryTransform(machine.get(), (AttributeSensor<?>)sensor);
+                    } catch (Exception e) {
+                        // TODO Avoid repeated logging
+                        Exceptions.propagateIfFatal(e);
+                        LOG.warn("Problem transforming sensor "+sensor+" of "+entity, e);
+                    }
+                }
+            }
+        }
+    }
+
+    protected void tryTransform(AttributeSensor<?> sensor) {
+        if (!isRunning()) {
+            return;
+        }
+        Maybe<MachineLocation> machine = getMachine();
+        if (machine.isAbsent()) {
+            return;
+        }
+        tryTransform(machine.get(), sensor);
+    }
+    
+    protected void tryTransform(MachineLocation machine, AttributeSensor<?> sensor) {
+        Object sensorVal = entity.sensors().get(sensor);
+        if (sensorVal == null) {
+            return;
+        }
+        Maybe<String> newVal = transformVal(machine, sensor, sensorVal);
+        if (newVal.isAbsent()) {
+            return;
+        }
+        AttributeSensor<String> mappedSensor = Sensors.newStringSensor(sensorNameConverter.apply(sensor.getName()));
+        if (newVal.get().equals(entity.sensors().get(mappedSensor))) {
+            // ignore duplicate
+            return;
+        }
+        LOG.debug("{} publishing value {} (original sensor value {}) for mapped sensor {}, of entity {}", 
+                new Object[] {this, newVal.get(), sensorVal, mappedSensor, entity});
+        entity.sensors().set(mappedSensor, newVal.get());
+    }
+    
+    protected Maybe<String> transformVal(MachineLocation machine, AttributeSensor<?> sensor, Object sensorVal) {
+        if (sensorVal == null) {
+            return Maybe.absent();
+        } else if (isPort(sensorVal)) {
+            int port = toInteger(sensorVal);
+            return transformPort(entity, machine, port);
+        } else if (isUri(sensorVal)) {
+            return transformUri(entity, machine, sensorVal.toString());
+        } else if (isHostAndPort(sensorVal)) {
+            return transformHostAndPort(entity, machine, sensorVal.toString());
+        } else {
+            // no-op; unrecognised type
+            return Maybe.absent();
+        }
+    }
+
+    protected boolean isUri(Object sensorVal) {
+        if (sensorVal instanceof URI || sensorVal instanceof URL) {
+            return true;
+        }
+        try {
+            URI uri = new URI(sensorVal.toString());
+            return uri.getScheme() != null;
+        } catch (URISyntaxException e) {
+            return false;
+        }
+    }
+
+    protected boolean isPort(Object sensorVal) {
+        if (sensorVal instanceof Integer || sensorVal instanceof Long) {
+            return Networking.isPortValid(((Number)sensorVal).intValue());
+        } else if (sensorVal instanceof CharSequence) {
+            return sensorVal.toString().trim().matches("[0-9]+");
```

Comment:
Is it not worth doing the isPortValid check here too? i.e. convert to number and then same as line 258